Implement avg_pool3d for MPS backend

donghao1393 opened the pull request:
This PR implements the avg_pool3d operation for the MPS backend using a custom Metal shader. This will allow users with Apple Silicon GPUs to use 3D average pooling operations without falling back to CPU.

## Implementation Details

The implementation includes:

1. A custom Metal shader for 3D average pooling
2. C++ interface to integrate with PyTorch
3. Support for both forward and backward passes
4. Comprehensive test cases
5. macOS version compatibility check (requires macOS 13.2+)
6. Special case handling for Long data type with divisor_override

## Development Challenges and Solutions

During the development process, we encountered several challenges:

1. **Metal Shader Compilation**: Initially, we faced issues with Metal shader compilation due to missing Xcode tools. We resolved this by ensuring proper Xcode installation and configuration.

2. **Command Buffer Conflicts**: When processing non-contiguous tensors, we encountered Metal command buffer errors with the message: 'A command encoder is already encoding to this command buffer'. This occurred because:
   - Metal requires that only one command encoder can be active for a command buffer at a time
   - When processing non-contiguous tensors, the tensor conversion operations were creating their own command encoders without properly ending previous ones
   - The error would occur when our code tried to create a new command encoder while another was still active

   We solved this by:
   - Adding explicit MPS stream synchronization before and after our operations using mpsStream->synchronize()
   - Creating a separate code path for non-contiguous tensors that first converts them to contiguous format
   - Ensuring proper command encoder lifecycle management by ending encoding before synchronizing the stream

3. **Version Compatibility**: We added explicit checks to ensure the implementation only runs on macOS 13.2 or newer, as earlier versions may not support all the required Metal features:
   ```cpp
   TORCH_CHECK(is_macos_13_or_newer(MacOSVersion::MACOS_VER_13_2_PLUS),
              "avg_pool3d is only supported on MPS for MacOS_13_2 or newer");
   ```

4. **Special Case Handling**: For certain edge cases (Long data type with divisor_override), we implemented a CPU fallback as MPS doesn't support these combinations efficiently.

5. **Improved Error Handling**: We added comprehensive dimension and data type checks to provide better error messages and ensure correct usage:
   - Checking input and output tensor dimensions
   - Verifying data type compatibility
   - Validating parameter values

## Implementation Approach

We chose to implement a custom Metal shader rather than using multiple 2D pooling operations or other approaches because:

1. It provides better performance for 3D data
2. It allows for more precise control over the pooling operation
3. It's consistent with how other 3D operations are implemented in PyTorch

## Alternative Approaches Considered

1. **Multiple 2D Pooling Operations**: We initially considered implementing avg_pool3d using multiple avg_pool2d operations, but this would have been less efficient and more complex to maintain.

2. **Using MPSCNNPooling**: We explored using the built-in Metal Performance Shaders for pooling, but they don't directly support 3D pooling operations.

3. **CPU Fallback**: The simplest approach would have been to fall back to CPU implementation, but this would have defeated the purpose of MPS acceleration.

## Update History

**April 19, 2025:**
Fixed issues with Metal command buffer handling when processing non-contiguous tensors. The solution ensures proper synchronization of MPS streams and correct handling of command encoders, avoiding the 'A command encoder is already encoding to this command buffer' error.

**April 20, 2025:**
- Marked PR as ready for review after comprehensive testing and verification
- Fixed linting issues and improved documentation
- Added comprehensive dimension and data type checks for better error handling
- Verified compatibility with various use cases, including those with the .out variant

This addresses issue #141287.

Fixes #151741 #141044

donghao1393 committed:
feat: Implement avg_pool3d for MPS backend

This commit implements the avg_pool3d operation for the MPS backend using a custom Metal shader.
The implementation includes:
1. A custom Metal shader for 3D average pooling
2. C++ interface to integrate with PyTorch
3. Support for both forward and backward passes
4. Comprehensive test cases

This addresses issue #141287.

donghao1393 committed:
ci: Add GitHub Actions workflow for MPS testing

donghao1393 committed:
ci: Update workflow name for clarity

donghao1393 committed:
ci: Update actions/upload-artifact to v4

donghao1393 committed:
ci: Update GitHub Actions to latest versions

donghao1393 committed:
ci: Switch back to high-performance runner

donghao1393 committed:
ci: Switch to standard runner for better cost-effectiveness

donghao1393 committed:
fix: Fix namespace and function call issues in avg_pool3d MPS implementation

donghao1393 committed:
trigger: Run GitHub Actions workflow

donghao1393 committed:
fix: Update avg_pool3d MPS implementation with structured kernels

donghao1393 committed:
trigger: Run GitHub Actions workflow

malfet commented on .github/workflows/build-test-mps.yml:
```diff
@@ -0,0 +1,50 @@
+name: Build and Test MPS Implementation
```
Why do you need this file?

donghao1393 replied:
@malfet Thanks for your review! You're right, we don't need the GitHub Actions workflow file anymore. We've found that local builds are much faster and more efficient for testing MPS implementations. I'll remove this file in the next update.

donghao1393 committed:
fix: Fix dispatch registration and string formatting in avg_pool3d MPS implementation

donghao1393 committed:
trigger: Run GitHub Actions workflow

donghao1393 committed:
fix: Update avg_pool3d MPS implementation to match kernel function signatures

donghao1393 committed:
fix: Update parameter names in avg_pool3d MPS implementation to match CPU implementation

donghao1393 committed:
fix: Update avg_pool3d_backward_kernel_impl parameter names to match Pool.h function type definition

donghao1393 committed:
fix: Use REGISTER_MPS_DISPATCH instead of REGISTER_DISPATCH for MPS implementation

donghao1393 committed:
fix: Use REGISTER_ARCH_DISPATCH instead of REGISTER_MPS_DISPATCH for MPS implementation

donghao1393 committed:
fix: Revert to REGISTER_MPS_DISPATCH for MPS implementation

donghao1393 committed:
fix: Use TORCH_IMPL_FUNC instead of kernel dispatch for MPS implementation

donghao1393 committed:
fix: Implement mps_avg_pool3d and mps_avg_pool3d_backward functions

donghao1393 committed:
fix: Implement avg_pool3d_out_mps and avg_pool3d_backward_out_mps functions

donghao1393 committed:
fix: Add static keyword to avg_pool3d_out_mps and avg_pool3d_backward_out_mps functions

donghao1393 committed:
fix: Use TORCH_IMPL_FUNC macro for avg_pool3d_out_mps and avg_pool3d_backward_out_mps functions

donghao1393 committed:
chore: Remove GitHub Actions workflow file

donghao1393 committed:
fix: Disable bfloat16 support in Pooling3D.metal for compatibility

donghao1393 committed:
fix: Use float32 instead of float in avg_pool3d test

donghao1393 committed:
fix: Explicitly handle only float and half types in avg_pool3d MPS implementation

donghao1393 committed:
fix: Force using metal_basic section to avoid issues with metal_bfloat

donghao1393 committed:
fix: Fallback to CPU implementation for avg_pool3d on MPS

donghao1393 committed:
fix: Restore MPS implementation for avg_pool3d after installing Xcode

donghao1393 committed:
fix: Correct avg_pool3d Metal shader implementation to match CPU behavior

donghao1393 committed:
fix: Correct pool size calculation in avg_pool3d Metal shader

donghao1393 committed:
fix: Use contiguous tensors in avg_pool3d MPS implementation

donghao1393 committed:
fix: Improve handling of non-contiguous tensors in avg_pool3d MPS implementation

donghao1393 committed:
fix: Improve Metal command buffer handling in avg_pool3d MPS implementation

donghao1393 committed:
fix: Fix command buffer handling in avg_pool3d MPS implementation

donghao1393 commented:
@malfet Thanks for your review! You're right, we don't need the GitHub Actions workflow file anymore. We've found that local builds are much faster and more efficient for testing MPS implementations. I'll remove this file in the next update.

donghao1393 commented:
@malfet Thanks for your review! You're right, we don't need the GitHub Actions workflow file anymore. We've found that local builds are much faster and more efficient for testing MPS implementations. 

We've updated the PR with the following improvements:

1. Fixed issues with Metal command buffer handling when processing non-contiguous tensors
2. Added proper macOS version compatibility check (requires macOS 13.2+)
3. Added special case handling for Long data type with divisor_override
4. Improved error handling and documentation

The implementation now passes all tests, including those with non-contiguous tensors.

donghao1393 committed:
Fix line length in test_mps.py

donghao1393 committed:
Improve avg_pool3d MPS implementation with better error handling and dimension checks

kulinseth commented on test/test_mps.py:
```diff
@@ -625,7 +625,7 @@ def mps_ops_modifier(ops):
         'nn.functional.max_unpool2dgrad': None,
         'nn.functional.max_unpool3dgrad': None,
-        'nn.functional.avg_pool3d': None,
+        # 'nn.functional.avg_pool3d': None,  # Now implemented
```
Cleanup.

donghao1393 replied:
cleaned up.

kulinseth commented on aten/src/ATen/native/mps/kernels/Pooling3D.metal:
```diff
@@ -0,0 +1,259 @@
+    constant int& stride_depth [[buffer(13)]],
+    constant int& stride_height [[buffer(14)]],
+    constant int& stride_width [[buffer(15)]],
+    constant int& padding_depth [[buffer(16)]],
```
You can use packed_int4 to pass the args easily

donghao1393 replied:
Used packed_int4 for more efficient parameter passing in avg_pool3d MPS implementation

donghao1393 replied:
https://github.com/pytorch/pytorch/pull/151742/commits/029a71255a741a0a84a360848abd871d5f1f2fa1

kulinseth commented on aten/src/ATen/native/mps/operations/Pooling.mm:
```diff
@@ -541,4 +548,354 @@ Tensor mps_max_pool2d_backward(const Tensor& grad_output,
+
+  // Synchronize the MPS stream
+  // This will commit the command buffer and wait for it to complete
+  mpsStream->synchronize(SyncType::COMMIT_AND_WAIT);
```
There is commit and wait at the beginning & end of this function. Not part of this checkin, but this should be cleaned out. We should be able to commit and continue bunch of work to GPU and not insert waits unless we need the results to be materialized by the compute streams.

donghao1393 replied:
https://github.com/pytorch/pytorch/pull/151742/commits/c793277fe65a5cde20345642e3a66d4987b5129c

kulinseth reviewed (commented):
Thanks for the change, overall looks good.

donghao1393 committed:
fix: Fix linting issues in avg_pool3d MPS implementation

donghao1393 committed:
fix: Remove commented out code for avg_pool3d in test_mps.py

donghao1393 committed:
perf: Use packed_int4 for more efficient parameter passing in avg_pool3d MPS implementation

donghao1393 committed:
perf: Reduce MPS stream synchronization in avg_pool3d for better performance

donghao1393 committed:
perf: Use Metal's built-in packed_int4 type for more efficient parameter passing and reduce synchronization

donghao1393 commented:
@kulinseth ready for review

donghao1393 committed:
fix: Fix formatting issues in Pooling.mm

donghao1393 committed:
fix: Fix formatting issues in Pooling3D.metal

donghao1393 commented:
force push to fix the conflict by rebase.
@kulinseth ready for review

donghao1393 committed:
Merge remote-tracking branch 'upstream/main' into feature/mps-avg-pool3d

donghao1393 commented:
new conflict arrived.

- reset back to https://github.com/pytorch/pytorch/commit/3adcc6802824cbee4e1a5d82bc9d5852e56f14f6 and force push.
- then fix the conflict, merge and push

cchance27 commented:
is this planned to be pulled into main pytorch in future? great to see more kernels for MPS being added

helloyokoy commented:
I hope to merge in as soon as possible， thanks
